### ISSUE TITLE
Encoding of Raw Data from KeyChain

### ISSUE THREAD

**Issue by rustymagnet3000** [NitinJami/keychaineditor#10]
I love this tool.  What a find.

I wanted to obtain all of the raw data stored in KeyChain for my app.  I ran:

./keychaineditor --action dump

I used the new build that works on iOS9.0.2 devices.  Everything worked. Or so I thought..

I was only interested in values stored in Generic Passwords.  Most fields returned were nice Base64 encoded.  I could convert them to Ascii to check the plaintext.  Perfect.

But where the fields  contained Application level Ciphertext (i.e. a HMAC key that was encrypted at rest over and above Keychain); KeyChainEditor appears to attempt to convert the ciphertext to Base64 and something goes wrong.  My conclusion is based on unexpected Backslash () characters in some of the fields.

An example - look at the Backslash (always next to a forward slash?. Don't get confused by the V character):

"Data" : "Dq+fQQ4BC7kJJ2yDb3ewsZUhh\/4teLa4enQQjf6pce07ETw+T0cDMHU14yDnYVLk"

**Comment by rustymagnet3000** [comment]
@NitinJami this is great stuff.  Thanks for putting in the effort to share this tool.

**Comment by hubert3** [comment]
Is there actually any error from Keychaineditor? It's quite possible that this is simply the payload your iOS app is storing in keychain, and it's using a Base64 representation with backslash escaped / characters.

**Comment by rustymagnet3000** [comment]
HI Hubert,  I had the same thought. But I checked - all of the Strings that are encrypted use NSUTF8StringEncoding

string1.dataUsingEncoding(NSUTF8StringEncoding)

**Comment by NitinJami** [comment]
Thanks @youdy3000 for trying out, and especially finding out bugs. This is actually a problem and I knew I would run into problems. Right now, I b64 all the values of ``Data'', irrespective of what form they are, in order for idb (http://www.idbtool.com/) to parse it properly. This is so that no data is lost between applications. 

However, a more intelligent way of doing this would be to first identify what is the encoding of data, and then do the necessary operation. For example, if the data value is already a b64 blob then I do not re-b64 it (which I am currently doing), and then deal with corner cases like the one's you pointed out.

Any suggestions here are welcome :)

**Comment by NitinJami** [comment]
Hey @youdy3000, I am unable to reproduce this. Can you give out some more details? specific examples, if possible. Thanks.

**Comment by rustymagnet3000** [comment]
Hey @NitinJami, good tip on idbtool.  I hope the below helps.  A couple of examples from KeyChainEditor:

Sunny Day scenario - 
  "54" : {
    "Protection" : "kSecAttrAccessibleAfterFirstUnlock",
    "Account" : "SomeField",
    "Data" : "WUVT"
  }

Convert  this into Ascii and you see the plaintext "YES" response.  

Rainy Day scenario - 

  "60" : {
    "Protection" : "kSecAttrAccessibleAfterFirstUnlock",
    "Account" : "SomeAppEncryptedField", 
    "Data" : "DxwWsLViRyve6tSC01EMmaVZFUX3njDLXR257YKVlrVyPbVWVMO6ndw1\/TqMDJ18"
}

Look in the Rainy Day scenario for the blackslash \ and the forwardslash /
The data field is a random (device created) string.  The string is AES\* encrypted and 
dataUsingEncoding:NSUTF8StringEncoding

In summary, At rest in keychain, the string is ciphertext and not plaintext. 

*ecb mode (no initialization vector) with a 256 bit key.

**Comment by NitinJami** [comment]
So, how are you decoding the above b64 blob? Are you using the openssl command in terminal. I can see that openssl command does not decode the above properly. However, when you use decoder in burpsuite, it decodes fine, and i bet idb will decode fine as well. (i now have your key ;-) )

I agree there is something weird going on only when using openssl command. I did play around with b64 encoding/decoding using Swift, and they tend to be compatible with openssl. Maybe something changed in Swift?

**Comment by NitinJami** [comment]
Closing this issue as there are no further comments.
